tensorflow/c/experimental/gradients/math_grad.cc

```diff
@@ -289,6 +292,118 @@ class MulGradientFunction : public GradientFunction {
   vector<AbstractTensorHandle*> forward_inputs;
 };
 
+class Log1pGradientFunction : public GradientFunction {
+ public:
+  explicit Log1pGradientFunction(vector<AbstractTensorHandle*> f_inputs)
+      : forward_inputs(f_inputs) {}
+
+  Status Compute(Context* ctx, const IncomingGradients& grad_inputs,
+                 vector<AbstractTensorHandle*>* grad_outputs) override {
+    // TODO(vnvo2409): Add control dependency
+    /* Given upstream grad U and a Log1p op: Y = log(1 + X), the gradients are:
+     *
+     *    dX = U / (1 + X)
+     *
+     */
+
+    AbstractTensorHandle* upstream_grad = grad_inputs[0];
+    AbstractTensorHandle* X = forward_inputs[0];
+
+    grad_outputs->resize(1);
+    vector<AbstractTensorHandle*> temp_outputs(1);
+
+    // Calculate conjugate of X
+    std::string name = "Conj_Log1p_Grad_X";
+    TF_RETURN_IF_ERROR(
+        Conj(ctx->ctx, {X}, absl::MakeSpan(temp_outputs), name.c_str()));
+
+    AbstractTensorHandle* Conj_X = temp_outputs[0];
+
+    // Creates Ones
+    name = "OnesLike_Log1p_Grad_X";
+    TF_RETURN_IF_ERROR(OnesLike(ctx->ctx, {Conj_X},
+                                absl::MakeSpan(temp_outputs), name.c_str()));
+
+    AbstractTensorHandle* Ones_X = temp_outputs[0];
+
+    name = "Add_Log1p_Grad_X";
+    // Calculate 1 + Conj(X)
+    TF_RETURN_IF_ERROR(Add(ctx->ctx, {Ones_X, Conj_X},
+                           absl::MakeSpan(temp_outputs), name.c_str()));
+
+    AbstractTensorHandle* Conj_XP1 = temp_outputs[0];
+
+    name = "DivNoNan_Log1p_Grad_X";
+    // Calculate U / (1 + Conj(X))
+    TF_RETURN_IF_ERROR(DivNoNan(ctx->ctx, {upstream_grad, Conj_XP1},
```

Comment:
`DivNoNan` basically eats up an error. I am not sure if doing that silently is a good idea. Maybe let's use `grad * reciprocal(1 + conj(x))` or `div(grad, 1 + conj(x))` here.

Reply:
Done